pkg/proxy/kafka.go

```diff
@@ -106,115 +106,113 @@ func createKafkaRedirect(r *Redirect, conf kafkaConfiguration) (RedirectImplemen
 
-func (k *kafkaRedirect) canAccess(req *kafka.RequestMessage, numIdentity identityPkg.NumericIdentity) bool {
-	var identity *identityPkg.Identity
+// canAccess determines if the kafka message req sent by identity is allowed to
+// be forwarded according to the rules configured on kafkaRedirect
+func (k *kafkaRedirect) canAccess(req *kafka.RequestMessage, srcIdentity identity.NumericIdentity) bool {
+	var id *identity.Identity
 
-	if numIdentity != 0 {
-		identity = identityPkg.LookupIdentityByID(numIdentity)
-		if identity == nil {
+	if srcIdentity != 0 {
+		id = identity.LookupIdentityByID(srcIdentity)
+		if id == nil {
 			log.WithFields(logrus.Fields{
 				logfields.Request:  req.String(),
-				logfields.Identity: numIdentity,
+				logfields.Identity: srcIdentity,
 			}).Warn("Unable to resolve identity to labels")
 		}
 	}
 
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.Request:  req.String(),
+		logfields.Identity: id,
+	})
+
 	k.redirect.mutex.RLock()
-	rules := k.redirect.rules.GetRelevantRules(identity)
+	rules := k.redirect.rules.GetRelevantRules(id)
 	k.redirect.mutex.RUnlock()
 
 	if rules.Kafka == nil {
-		flowdebug.Log(log.WithField(logfields.Request, req.String()),
-			"No Kafka rules loaded, rejecting")
+		flowdebug.Log(scopedLog, "No Kafka rules matching identity, rejecting")
 		return false
 	}
 
 	b, err := json.Marshal(rules.Kafka)
 	if err != nil {
-		flowdebug.Log(log.WithError(err).WithField(logfields.Request, req.String()),
-			"Error marshalling kafka rules to apply")
+		flowdebug.Log(scopedLog, "Error marshalling kafka rules to apply")
 		return false
 	} else {
-		flowdebug.Log(log.WithFields(logrus.Fields{
-			logfields.Request: req.String(),
-			"rule":            string(b),
-		}), "Applying rule")
+		flowdebug.Log(scopedLog.WithField("rule", string(b)), "Applying rule")
 	}
 
 	return req.MatchesRule(rules.Kafka)
 }
 
 // kafkaLogRecord wraps an accesslog.LogRecord so that we can define methods with a receiver
 type kafkaLogRecord struct {
-	accesslog.LogRecord
-	redirect *Redirect
-	req      *kafka.RequestMessage
+	logger.LogRecord
+	topics []string
 }
 
-func (k *kafkaRedirect) newKafkaLogRecord(req *kafka.RequestMessage) kafkaLogRecord {
-	record := kafkaLogRecord{
-		LogRecord: req.GetLogRecord(),
-		req:       req,
-		redirect:  k.redirect,
+func apiKeyToString(apiKey int16) string {
+	if key, ok := api.KafkaReverseAPIKeyMap[apiKey]; ok {
+		return key
 	}
+	return fmt.Sprintf("%d", apiKey)
+}
 
-	if k.redirect.ingress {
-		record.ObservationPoint = accesslog.Ingress
-	} else {
-		record.ObservationPoint = accesslog.Egress
+func (k *kafkaRedirect) newLogRecordFromRequest(req *kafka.RequestMessage) kafkaLogRecord {
+	return kafkaLogRecord{
+		LogRecord: logger.NewLogRecord(k.redirect, accesslog.TypeRequest,
+			logger.LogTags.Kafka(&accesslog.LogRecordKafka{
+				APIVersion:    req.GetVersion(),
+				APIKey:        apiKeyToString(req.GetAPIKey()),
+				CorrelationID: int32(req.GetCorrelationID()),
```

Comment:
Should we also be filling `ClientID` if present in the request?

Reply:
I'm not opposed to adding the ClientID to the log record but it's outside of the scope of this PR.